File: src/gui/src/app/classes/events/core/unsupported-version-event.ts
import { BaseEvent, EventLogLevel } from '@app/interfaces/events';
import { UnsupportedVersionEvent as ProtoUnsupportedVersionEvent } from '@gen/es/fme/v1/events_pb';
import { ListEventsResponse } from '@gen/es/fme/v1/fme_service_pb';
import { docsLinks } from '@app/constants/external-links';

export class UnsupportedVersionEvent implements BaseEvent {
    logLevel = EventLogLevel.Warning;

    constructor(public newVersion: string, public currentVersion: string, public releaseNotes: string[]) {
    }

    get logMessage(): string {
        return `Your current version is no longer supported. Current version: ${this.currentVersion}, Newest supported version: ${this.newVersion}. Visit ${docsLinks.GITHUB_REPO} to download`;
    }

    static fromProtobuf(event: ListEventsResponse): UnsupportedVersionEvent {
        const evt = event.event.value as ProtoUnsupportedVersionEvent;

        return new UnsupportedVersionEvent(
            evt.newSupportedVersion,
            evt.currentVersion,
            evt.releaseNotes,
        );
    }
}

File: src/cli/types/eventtypes/unsupported_version_event.go
package eventtypes

import (
	"fmt"

	"github.com/awslabs/filemoverexpress/constants"
	"github.com/awslabs/filemoverexpress/logger"
	"github.com/awslabs/filemoverexpress/types/pbtypes/fme/v1"
)

type UnsupportedVersionEvent struct {
	NewSupportedVersion string
	CurrentVersion      string
	ReleaseNotes        []string
}

func (nva *UnsupportedVersionEvent) String() string {
	msg := "Your current version is no longer supported. Current version: %s, Newest supported version: %s. Visit %s to download"
	return fmt.Sprintf(
		msg,
		nva.CurrentVersion,
		nva.NewSupportedVersion,
		constants.MarketingPageUrl,
	)
}

func (*UnsupportedVersionEvent) Type() MessageFlags {
	return UnsupportedVersionEventType
}

func (*UnsupportedVersionEvent) Priority() logger.LogLevel {
	return UnsupportedVersionEventPriority
}

func (nva *UnsupportedVersionEvent) ToProtobuf() (fmev1.PbEvent, fmev1.EventType) {
	pbEvent := &fmev1.UnsupportedVersionEvent{
		NewSupportedVersion: nva.NewSupportedVersion,
		CurrentVersion:      nva.CurrentVersion,
		ReleaseNotes:        nva.ReleaseNotes,
	}
	msgEvent := fmev1.ListEventsResponse_UnsupportedVersionEvent{UnsupportedVersionEvent: pbEvent}
	return &msgEvent, fmev1.EventType_EVENT_TYPE_UNSUPPORTED_VERSION_EVENT_TYPE
}
